CRM/MembersOnlyEvent/BAO/EventGroup.php

```diff
@@ -130,33 +128,70 @@ public static function getContactGroupIDs($contactID) {
     if (self::$contactGroupIDs !== NULL) {
       return self::$contactGroupIDs;
     }
-    self::$contactGroupIDs = [];
 
+    self::$contactGroupIDs = array_merge(
+      self::getContactNormalGroupIds($contactID),
+      self::getContactSmartGroupIds($contactID)
+    );
+
+    return self::$contactGroupIDs;
+  }
+
+  /**
+   * Gets the normal groups for the specified contact.
+   *
+   * @param int $contactID
+   *
+   * @return array
+   *   List of group ids or empty array if nothing found
+   */
+  public static function getContactNormalGroupIds($contactID) {
```

Comment:
this does not need to be public, making something public just to be able to test it is not a good idea, you should only test the class public interface.

this also applies to:

- getContactSmartGroupIds()
- getContactGroupIDs()

and thus, you tests in tests/phpunit/CRM/MembersOnlyEvent/BAO/EventGroupTest.php should test only getContactAllowedGroups.

No you can write multiple tests for this method, maybe such as:

- testIfGetContactAllowedGroupsReturnsNothingIfNoAllowedGroupsAreConfigured
- testIfGetContactAllowedGroupsReturnsNothingIfAllowedGroupsAreConfiguredButTheContactIsNotPartOfAnyOfThem
- testIfGetContactAllowedGroupsWorkWithSmartGroups
- testIfGetContactAllowedGroupsWorkWithNormalGroups

Reply:
Updated except for updating the `getContactGroupIDs` method to be private because I use it in the file [MembersOnlyEventAccess.php#L212](https://github.com/compucorp/uk.co.compucorp.membersonlyevent/blob/2541f0ae6f7813002da413371a444fae3532bf5e/CRM/MembersOnlyEvent/Service/MembersOnlyEventAccess.php#L212).